### PR TITLE
Issue #3464 - Split SslContextFactory into Client and Server

### PR DESCRIPTION
**Pull request by sbordet**
Issue #3464.

Introduced SslContextFactory subclasses Client and Server.
Replaced all usages of SslContextFactory with either Client or Server
as required.

Refactored configuration checking so that warnings are not emitted
when non necessary.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

**Comment by sbordet** [comment]
Large PR but the main changes are in `SslContextFactory`, in 2 places:

1. the introduction of `Client` and `Server` subclasses
1. the `checkConfiguration()` method

All the rest is just converting all our usages of `SslContextFactory` to either `Client` or `Server` as required, in examples and tests.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java [hunk 1]
```diff
@@ -1953,4 +1987,56 @@ public X509 getX509()
...
+        public Client(boolean trustAll)
+        {
+            super(trustAll);
+        }
+
+        @Override
+        protected void checkConfiguration()
+        {
+            checkTrustAll();
+            checkEndPointIdentificationAlgorithm();
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: Should we warn if someone sets up wantClientAuth or needClientAuth to true in the Client mode?

**Comment by sbordet** [comment]
@joakime no, the server can use `EndpointIdentificationAlgorithm="HTTPS"` if they wish so (just by default they don't)

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java (hunk 1 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: Don't know, it's not really an "insecure" configuration, just one that has no effect.

Eventually in Jetty 10 `SslContextFactory` will be `abstract` and those methods only present in `Server`.

Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: Good point, do we have test cases still around that keep the old `new SslContextFactory()` usage?

**Comment by joakime** [comment]
If someone copied / pasted the `jetty-ssl.xml` to their jetty-base/etc/ and modified it from say Jetty 9.4.12.
Would it still work with these changes?

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java (hunk 1 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: No, now we do not have code that uses `SslContextFactory` directly, it's all either `Client` or `Server`.

**Comment by sbordet** [comment]
It will still work in 9.4.x, but won't work in 10.0.x.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java [hunk 2]
```diff
@@ -1953,4 +1987,56 @@ public X509 getX509()
             return _x509;
         }
     }
+
+    public static class Client extends SslContextFactory
+    {
+        public Client()
+        {
+            this(false);
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: this constructor is deprecated.  So we need a new non deprecated protected constructor that can be used by both Client and Server.  So make the current `private SslContextFactory(boolean trustAll, String keyStorePath)`  a protected constructor and call it directly.
The old public constructors can then log warnings!

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java [hunk 3]
```diff
@@ -1953,4 +1987,56 @@ public X509 getX509()
...
+    {
+        public Server()
+        {
+            setEndpointIdentificationAlgorithm(null);
+        }
+
+        @Override
+        public boolean getWantClientAuth()
+        {
+            return super.getWantClientAuth();
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: these are also deprecated methods!  Would be better to directly access the field.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java (hunk 2 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: The constructor is not deprecated. It is calling a new `Client(boolean)` constructor.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java (hunk 3 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: It is not better to access the field. When merging to 10, I will take care of _moving_ the fields here.
So, for now, it stays like that, in 10 will be cleaned up properly.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java [hunk 4]
```diff
@@ -1761,10 +1791,14 @@ public SSLParameters customize(SSLParameters sslParams)
             sslParams.setCipherSuites(_selectedCipherSuites);
         if (_selectedProtocols != null)
             sslParams.setProtocols(_selectedProtocols);
-        if (getWantClientAuth())
-            sslParams.setWantClientAuth(true);
-        if (getNeedClientAuth())
-            sslParams.setNeedClientAuth(true);
+        if (this instanceof Server)
```
Comment (bentmann) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: This unfortunately breaks compatibility with code originally written against earlier Jetty versions:
Such code instantiates an instance of `SslContextFactory` (and not a subclass thereof) which doesn't satisfies this type check and as a result no longer has the two client auth options set for its server.

Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java]: Doh, good catch! Thanks!

**Pull request merged**